File: leetcode/continuous-subarray-sum.py
# 523. Continuous Subarray Sum
# 🟠 Medium
#
# https://leetcode.com/problems/continuous-subarray-sum/
#
# Tags: Array - Hash Table - Math - Prefix Sum

import timeit
from typing import List


# Compute the prefix sums then iterate over each combination of two
# indexes checking if the sum of values between them is a multiple of k.
#
# Time complexity: O(n^2) - We iterate over all the combinations of two
# indexes.
# Space complexity: O(n) - The prefix sum array has the same length as
# the input.
#
# This solution fails with Time Limit Exceeded.
class Naive:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        # Compute an array of prefix sums.
        pre = [nums[0]] + ([0] * (len(nums) - 1))
        for i in range(1, len(nums)):
            pre[i] = nums[i] + pre[i - 1]
        # Iterate over all pairs of two indexes.
        for i in range(len(nums) - 1):
            for j in range(i + 1, len(nums)):
                range_sum = pre[j]
                if i > 0:
                    range_sum -= pre[i - 1]
                if range_sum % k == 0:
                    return True
        return False


# Iterate over the input computing the cumulative sum up to the current
# index, calculate sum[0:i] % k and store the result in a hashmap, if we
# ever find a repeated value, the sum of values between the indexes that
# resulted in the same remainder will be a multiple of k.
#
# Time complexity: O(n) - We iterate over the array once.
# Space complexity: O(k) - The hashmap holds the result of mod K
# operations so it can grow at most to size k.
#
# Runtime: 1563 ms, faster than 71.44%
# Memory Usage: 33.3 MB, less than 39.30%
class UseHashMap:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        # Use a hashmap. Initialize it with a zero to represent the sum
        # of values at the index left of 0.
        seen = {0: 0}
        current_sum = 0
        for i in range(len(nums)):
            current_sum += nums[i]
            rem = current_sum % k
            if rem % k not in seen:
                seen[rem] = i + 1
            # If we have seen this sum in a previous index.
            elif seen[rem] < i:
                return True
        return False


def test():
    executors = [
        Naive,
        UseHashMap,
    ]
    tests = [
        [[0], 1, False],
        [[10], 10, False],
        [[5, 10, 35], 10, True],
        [[23, 2, 4, 6, 7], 6, True],
        [[23, 2, 4, 6, 6], 7, True],
        [[23, 2, 6, 4, 7], 6, True],
        [[23, 2, 6, 4, 7], 13, False],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.checkSubarraySum(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/continuous-subarray-sum.rs
// 523. Continuous Subarray Sum
// 🟠 Medium
//
// https://leetcode.com/problems/continuous-subarray-sum/
//
// Tags: Array - Hash Table - Math - Prefix Sum

use std::collections::HashMap;

struct Solution;
impl Solution {
    /// Use the fact that for the subarray sum between i<>j to be a multiple of k, the mod k of
    /// both prefix sums up to i and j needs to be equal. There is a better explanation of why in
    /// the problem editorial. Initialize the `prefix_mod` to 0, keep a hash map of prefix mods
    /// seen, for each value, add the current value and mod it to get the prefix mod up to the
    /// current index, if we have seen that mod and the array between the indexes has at least
    /// length 2, we return true.
    ///
    /// Time complexity: O(n) - We visit each element in nums and do constant time work for each.
    /// Space complexity: O(n) - The HashMap of seen mods.
    ///
    /// Runtime 26 ms Beats 60%
    /// Memory 4.56 MB Beats 60%
    pub fn check_subarray_sum(nums: Vec<i32>, k: i32) -> bool {
        let mut pm = 0;
        let mut seen = HashMap::new();
        seen.insert(0, -1);
        for (i, num) in nums.into_iter().enumerate() {
            pm = (pm + num) % k;
            match seen.get(&pm) {
                Some(j) => {
                    if i as i32 - j > 1 {
                        return true;
                    }
                }
                None => {
                    seen.insert(pm, i as i32);
                }
            }
        }
        false
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![2, 4, 3], 6, true),
        (vec![23, 2, 4, 6, 7], 6, true),
        (vec![23, 2, 6, 4, 7], 6, true),
        (vec![23, 2, 6, 4, 7], 13, false),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::check_subarray_sum(t.0.clone(), t.1);
        if res == t.2 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.2, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
